### PR TITLE
Added support for Cardputer keyboard input.

### PR DESCRIPTION
**Pull request by joshua-beck-0908**
Added support for using the built-in keyboard for serial input to the M5Stack Card Computer board.
Some modifications were made to the serial input code to allow a per-board serial input source rather than just per port.
The scanner deinitialisation code was also modified to allow the never reset functionality on scanners in order prevent the keyboard from being deinitialised after a program finishes.

### CODE REVIEW

File: shared-module/keypad/__init__.c [hunk 1]
```diff
@@ -58,8 +58,14 @@ void keypad_tick(void) {
 }
 
 void keypad_reset(void) {
```
Comment (tannewt) [inline shared-module/keypad/__init__.c:60]: @dhalbert Couldn't we switch this to using a finalizer instead?

File: shared-module/keypad_demux/DemuxKeyMatrix.c [hunk 2]
```diff
@@ -146,3 +146,13 @@ static void demuxkeymatrix_scan_now(void *self_in, mp_obj_t timestamp) {
...
 }
+
+void demuxkeymatrix_never_reset(keypad_demux_demuxkeymatrix_obj_t *self) {
+    keypad_never_reset((keypad_scanner_obj_t *)self);
+    for (size_t row_addr = 0; row_addr < self->row_addr_digitalinouts->len; row_addr++) {
+        common_hal_digitalio_digitalinout_never_reset(self->row_addr_digitalinouts->items[row_addr]);
+    }
+    for (size_t column = 0; column < self->column_digitalinouts->len; column++) {
+        common_hal_digitalio_digitalinout_never_reset(self->column_digitalinouts->items[column]);
+    }
```
Comment (tannewt) [inline shared-module/keypad_demux/DemuxKeyMatrix.c:157]: You'll need this even if we switch keypads to finalizers because pins are still bulk reset.

**Review by tannewt — CHANGES_REQUESTED** [review]
Supervisor changes look good! A couple other things. You can add a private API to keypad for other C modules to register callbacks if you like. That way you could map immediately and raise ctrl-c if needed.

File: supervisor/shared/serial.c [hunk 3]
```diff
@@ -107,6 +107,29 @@ int console_uart_printf(const char *fmt, ...) {
...
+}
+
+MP_WEAK char board_serial_read(void) {
+    return -1;
+}
+
+MP_WEAK uint32_t board_serial_bytes_available(void) {
+    return false;
+}
+
```
Comment (RetiredWizard) [inline supervisor/shared/serial.c:127]: Should this be
    return 0;

Comment (dhalbert) [reply supervisor/shared/serial.c:127]: Due to #9141 being recently merged, yes, this should be `0`.

**Comment by RetiredWizard** [comment]
There aren't any artifacts for this (probably because of the rebasing conflict) but I manually merged the changes on a local build and it seems to be working well.

**Comment by joshua-beck-0908** [comment]
Fixed the incorrect prototype for board_serial_bytes_available.

**Comment by dhalbert** [comment]
Note the merge conflicts.

**Comment by joshua-beck-0908** [comment]
I had to add some of the `board_` prototypes back in hopefully that settles it!

File: shared-module/keypad/__init__.c (hunk 1 above)
Comment (dhalbert) [reply shared-module/keypad/__init__.c:60]: Certainly, but is there other stuff first? Is the PR blocked if it's not a finalizer yet?

Comment (joshua-beck-0908) [reply shared-module/keypad/__init__.c:60]: If I could get some advice on implementing one of these finalizers that would be great.

**Comment by joshua-beck-0908** [comment]
I just added in a little extra code to get the OPT modifier on the keyboard to produce a unique keycode when pressed with another key.
FN, Shift, Ctrl, Opt, and Alt... they sure have a lot of modifiers for one little keyboard.

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**